pkg/core/engine/engine.go

```diff
@@ -0,0 +1,473 @@
+				return nil, failure
+			}
+		}
+
+		if onSteakReady != nil {
+			onSteakReady(steak)
+		}
+
+		for _, topic := range taggedBEEF.Topics {
+			admit := steak[topic]
+			outputsConsumed := make([]*Output, 0, len(admit.CoinsToRetain))
+			outpointsConsumed := make([]*overlay.Outpoint, 0, len(admit.CoinsToRetain))
+			for _, vin := range admit.CoinsToRetain {
+				output := topicInputs[topic][vin]
+				if output == nil {
+					return nil, ErrMissingInput
+				}
+				outputsConsumed = append(outputsConsumed, output)
+				outpointsConsumed = append(outpointsConsumed, &output.Outpoint)
+				delete(topicInputs[topic], vin)
+			}
+			for vin, output := range topicInputs[topic] {
+				if err := e.deleteUTXODeep(ctx, output); err != nil {
+					return nil, err
+				}
+				admit.CoinsRemoved = append(admit.CoinsRemoved, uint32(vin))
+			}
+
+			newOutpoints := make([]*overlay.Outpoint, 0, len(admit.OutputsToAdmit))
+			for _, vout := range admit.OutputsToAdmit {
+				out := tx.Outputs[vout]
+				output := &Output{
+					Outpoint: overlay.Outpoint{
+						Txid:        *txid,
+						OutputIndex: uint32(vout),
+					},
+					Script:          out.LockingScript,
+					Satoshis:        out.Satoshis,
+					Topic:           topic,
+					OutputsConsumed: outpointsConsumed,
+					Beef:            taggedBEEF.Beef,
+				}
+				if err := e.Storage.InsertOutput(ctx, output); err != nil {
+					return nil, err
+				}
+				newOutpoints = append(newOutpoints, &output.Outpoint)
+				for _, l := range e.LookupServices {
+					if err := l.OutputAdded(ctx, output); err != nil {
+						return nil, err
+					}
+				}
+			}
+			for _, output := range outputsConsumed {
+				output.ConsumedBy = append(output.ConsumedBy, newOutpoints...)
+
+				if err := e.Storage.UpdateConsumedBy(ctx, &output.Outpoint, output.Topic, output.ConsumedBy); err != nil {
+					return nil, err
+				}
+			}
+			if err := e.Storage.InsertAppliedTransaction(ctx, &overlay.AppliedTransaction{
+				Txid:  txid,
+				Topic: topic,
+			}); err != nil {
+				return nil, err
+			}
+		}
+		if e.Advertiser == nil || mode == SubmitModeHistorical {
+			return steak, nil
+		}
+
+		//TODO: Implement SYNC
+
+		return steak, nil
+	}
+}
+
+func (e *Engine) Lookup(ctx context.Context, question *lookup.LookupQuestion) (*lookup.LookupAnswer, error) {
+	if l, ok := e.LookupServices[question.Service]; ok {
+		return nil, ErrUnknownTopic
+	} else if result, err := l.Lookup(ctx, question); err != nil {
+		return nil, err
+	} else if result.Type == lookup.AnswerTypeFreeform || result.Type == lookup.AnswerTypeOutputList {
+		return result, nil
+	} else {
+		hydratedOutputs := make([]*lookup.OutputListItem, 0, len(result.Outputs))
+		for _, formula := range result.Formulas {
+			if output, err := e.Storage.FindOutput(ctx, formula.Outpoint, nil, nil, true); err != nil {
+				return nil, err
+			} else if output != nil && output.Beef != nil {
+				if output, err := e.GetUTXOHistory(ctx, output, formula.Histoy, 0); err != nil {
```

Comment:
```suggestion
				if output, err := e.GetUTXOHistory(ctx, output, formula.Histoy, 0); err != nil {
```
here is shadowing of `output`